### PR TITLE
Add support for partitioned cookies to CookieStore.

### PR DESCRIPTION
**Pull request by DCtheTall**
***
<a href="https://pr-preview.s3.amazonaws.com/DCtheTall/cookie-store/pull/206.html" title="Last updated on Nov 3, 2021, 1:40 PM UTC (976425a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/cookie-store/206/e1a1cfa...DCtheTall:976425a.html" title="Last updated on Nov 3, 2021, 1:40 PM UTC (976425a)">Diff</a>

***
<a href="https://pr-preview.s3.amazonaws.com/DCtheTall/cookie-store/pull/206.html" title="Last updated on Aug 4, 2023, 9:29 PM UTC (28de9d0)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/cookie-store/206/c9a8c1d...DCtheTall:28de9d0.html" title="Last updated on Aug 4, 2023, 9:29 PM UTC (28de9d0)">Diff</a>

### CODE REVIEW

**Comment by DCtheTall** [comment]
This PR is a change to the CookieStore API spec to support partitioned cookies.

These changes could be used to support the [Partitioned attribute, a.k.a. CHIPS](https://github.com/WICG/CHIPS).

There is an [open PR](https://github.com/WICG/CHIPS/pull/24) in the CHIPS repository adding a note about these changes to the explainer.

File: index.bs
Comment (domenic) [inline index.bs:510]:
```suggestion
  boolean partitioned;
```

Comment (DCtheTall) [reply index.bs]: Done.

**Comment by DCtheTall** [comment]
One quick addendum: if user agents implement service worker partitioning ([ref](https://github.com/privacycg/storage-partitioning)) then the cookie's partition key should be the site of the worker's partition key.

If user agents don't implement storage partitioning, then partitioned cookies may need to be blocked in service workers. Otherwise, service workers could use partitioned cookies as a cross-site identifier. See [the CHIPS explainer](https://github.com/WICG/CHIPS#service-workers) for more detail.

**Comment by DCtheTall** [comment]
I put together a [WIP CL](https://chromium-review.googlesource.com/c/chromium/src/+/3212611) that implements these changes in Chromium.

The CL fails some web platform tests (for now 🤞), but you can get the general idea of how these changes would materialize from it.

**Comment by inexorabletash** [comment]
Wow, this fell off my radar. Is this just waiting for final review/merge, or is there further discussion that should happen first?

File: index.bs
Comment (inexorabletash) [reply index.bs:696]: ... and "current browsing context" isn't the right term. I think "current global object" ?

Comment (inexorabletash) [reply index.bs:696]: Also.. this is run after the cookie is set, which seems unusual. ISTM the partition key should be captured before "set a cookie" and passed into it.

File: explainer.md
Comment (inexorabletash) [inline explainer.md:376]: In the IDL the default isn't given for `CookieStoreDeleteOptions`

File: index.bs
Comment (inexorabletash) [inline index.bs:752]: "field" isn't the right terminology.  Also, per previous comments, if we can default this it simplifies the logic.

File: explainer.md
Comment (dmdabbs) [inline explainer.md:280]: CHIPS [does not mandate](https://github.com/privacycg/CHIPS#handling-older-or-incompatible-clients) use of the __Host-prefix, 

>Although it is not required, it is still recommended to still include the __Host- prefix. Even clients that do not recognize the Partitioned attribute still enforce the semantics of the __Host- prefix. 

so a request could present two cookies with the same name. Will `cookieStore.get()` fail or simply return the first of the two cookies? In that situation, would `cookieStore.getAll()` retrieve the two cookies?

Comment (inexorabletash) [reply explainer.md:280]: #221 is relevant here - there are existing situations where multiple cookies with the same name can be present, not a new situation with partitioned cookies.

Currently as spec'd and implemented in Chrome, the first cookie is returned.

File: explainer.md
Comment (DCtheTall) [reply explainer.md:317]: Done

File: explainer.md
Comment (DCtheTall) [reply explainer.md:357]: Done

File: explainer.md
Comment (DCtheTall) [reply explainer.md]: Changed the wording

File: index.bs
Comment (DCtheTall) [reply index.bs]: Removed mentioning of partition key in this section.

File: explainer.md
Comment (DCtheTall) [reply explainer.md]: Added default to IDL

File: index.bs
Comment (DCtheTall) [reply index.bs]: Done

File: explainer.md
Comment (DCtheTall) [reply explainer.md]: Done, forgot to resolve.

File: index.bs
Comment (DCtheTall) [reply index.bs]: I think the partition key can be up to RFC6265bis to define, since the CookieStore API is not actually concerned with partition key values and the cookie RFC will (some day) specify how to select which partitioned cookies to include in a request. So I do not think this spec needs to be concerned with that?

File: index.bs
Comment (DCtheTall) [reply index.bs]: Done.

File: index.bs
Comment (inexorabletash) [reply index.bs]: Yes... but https://wicg.github.io/cookie-store/#cookie-concept basically has "proxy" definitions for RFC6265bis terms, since I don't think we can link directly to them. We can add it later though.

File: index.bs
Comment (inexorabletash) [inline index.bs:704]: Since this defaults to true now, this step can be removed...

File: index.bs
Comment (inexorabletash) [inline index.bs:715]: ... and this can just be `|options|["{{CookieStoreDeleteOptions/partitioned}}"]` now.

File: index.bs
Comment (inexorabletash) [inline index.bs:772]: nit: the sentence lost the "and" and the trailing period.

**Review by inexorabletash — CHANGES_REQUESTED** [review]
Almost there! Just tiny nitpicks.

**Comment by inexorabletash** [comment]
BTW, I was playing with cookieStore/document.cookie interop and it took me a while to realize that with document.cookie "partitioned" is ignored unless "path" and "secure" are also specified, which are the default with cookieStore but of course not document.cookie. Do you think we should add an example in the "Modifying Cookies" section, or is this well known?

File: index.bs
Comment (DCtheTall) [reply index.bs]: Done.

File: index.bs
Comment (DCtheTall) [reply index.bs]: Done.

File: index.bs
Comment (DCtheTall) [reply index.bs]: Done.

**Comment by DCtheTall** [comment]
> BTW, I was playing with cookieStore/document.cookie interop and it took me a while to realize that with document.cookie "partitioned" is ignored unless "path" and "secure" are also specified

Tad correction, you only need the "secure" attribute set. The "path" attribute is not necessary.

> Do you think we should add an example in the "Modifying Cookies" section, or is this well known?

The Secure requirement is a well-known part of the proposal for the definition of Partitioned we have for RFC6265bis.

**Comment by inexorabletash** [comment]
> > BTW, I was playing with cookieStore/document.cookie interop and it took me a while to realize that with document.cookie "partitioned" is ignored unless "path" and "secure" are also specified
> 
> Tad correction, you only need the "secure" attribute set. The "path" attribute is not necessary 😄

Thanks for clarifying! https://source.chromium.org/chromium/chromium/src/+/main:net/cookies/canonical_cookie_unittest.cc;l=731 mislead me a bit I guess.

**Comment by DCtheTall** [comment]
> Thanks for clarifying! [source.chromium.org/chromium/chromium/src/+/main:net/cookies/canonical_cookie_unittest.cc;l=731](https://source.chromium.org/chromium/chromium/src/+/main:net/cookies/canonical_cookie_unittest.cc;l=731) mislead me a bit I guess.

Ah, this is because Path=/ is required for the __Host- prefix, which that cookie also uses. I can see why that is initially confusing, especially considering Partitioned was originally designed to require __Host- but we loosened it to only require Secure 😄

**Pull request merged**
